Donation added (for coss review no test) 1.0v

salikhdev opened the pull request.

salikhdev committed:
Donation is added (For cross review no test) 1.0v

az3l1t commented on src/main/java/faang/school/projectservice/controller/DonationController.java:
```diff
@@ -0,0 +1,35 @@
+    }
+
+    @GetMapping("/{id}")
+    public DonationDto getDonationById(@Valid @PathVariable Long id, @Valid @RequestParam Long userId) {
```
а зачем @Valid пишешь если ничего не валидируешь?

az3l1t commented on src/main/java/faang/school/projectservice/controller/DonationController.java:
```diff
@@ -0,0 +1,35 @@
+    }
+
+    @GetMapping("/all")
+    public List<DonationDto> getUserDonations(@Valid @RequestParam Long userId, @RequestBody DonationFilterDto filterDto) {
```
а зачем @Valid пишешь если ничего не валидируешь?

az3l1t commented on src/main/java/faang/school/projectservice/dto/donate/DonationCreateDto.java:
```diff
@@ -0,0 +1,17 @@
+        Long campaignId,
+        BigDecimal amount,
+        Currency currency,
+        LocalDateTime donationTime
```
Можешь тут в конструкторе сразу писать, что LocalDateTime = LocalDateTime.now()
меньше кода будет

salikhdev replied:
![image](https://github.com/user-attachments/assets/7cea959d-3fd1-4d5b-bd01-9596436c8ac3)

az3l1t replied:
немного не так) я имел в виду в конструкторе,
но я апрувнул. в целом можешь не менять, это просто на будущее
![image](https://github.com/user-attachments/assets/ca6e8f80-0d85-41e4-be20-7cea13547d19)

az3l1t commented on src/main/java/faang/school/projectservice/filter/donation/MaxAmountFilter.java:
```diff
@@ -0,0 +1,26 @@
+package faang.school.projectservice.filter.donation;
```
не вариант.
мы храним фильтры в другом месте, если напишешь тут, то опять будут проблемы

az3l1t commented on src/main/java/faang/school/projectservice/service/donate/DonationService.java:
```diff
@@ -0,0 +1,114 @@
+
+        if (!paymentResponse.status().equals("SUCCESS")) {
+            log.error("Payment Failed ! {}, {}", donationDto, paymentResponse);
+            throw new PaymentFailedException("Payment Failed !");
```
создал свой exception, надо добавить его сразу в GlobalExceptionHandler, чтобы 
нормально выводить определенную ошибку при запросе

az3l1t commented on src/main/java/faang/school/projectservice/service/donate/DonationService.java:
```diff
@@ -0,0 +1,114 @@
+            return paymentServiceClient.sendPayment(paymentRequest);
+        } catch (Exception e) {
+            log.error("Payment service not working ! {}", paymentRequest);
+            throw new PaymentServiceConnectException("Payment service not working !");
```
создал свой exception, надо добавить его сразу в GlobalExceptionHandler, чтобы
нормально выводить определенную ошибку при запросе

az3l1t commented on src/main/java/faang/school/projectservice/service/donate/DonationService.java:
```diff
@@ -0,0 +1,114 @@
+    }
+
+    private PaymentResponse paymentToDonate(DonationCreateDto donationDto) {
+        var paymentNumber = new Random().nextLong(MIN_PAYMENT_NUMBER, MAX_PAYMENT_NUMBER);
```
каждый запрос создаешь новый Random, подумай как сделать это эффективнее

az3l1t commented on src/main/java/faang/school/projectservice/service/donate/DonationService.java:
```diff
@@ -0,0 +1,114 @@
+        try {
+            return paymentServiceClient.sendPayment(paymentRequest);
+        } catch (Exception e) {
+            log.error("Payment service not working ! {}", paymentRequest);
```
вот тут я бы передал ошибку e в лог

az3l1t commented on src/main/java/faang/school/projectservice/service/donate/DonationService.java:
```diff
@@ -0,0 +1,114 @@
+
+        Campaign campaign = campaignService.getCampingById(donationDto.campaignId());
+
+        PaymentResponse paymentResponse = paymentToDonate(donationDto);
```
Если `paymentServiceClient.sendPayment()` выполняется успешно, но дальше код падает, транзакция отменится, но платёж уже уйдет. Это приведёт к потере данных, доната нет, а платеж есть допустим.
->
сделать сначала сохранение в БД, потом платёж (вообще было бы классно если бы был откат платежа)

az3l1t commented:
классно, что много логов сделал

snowmangoofficial commented on src/main/java/faang/school/projectservice/dto/donate/DonationCreateDto.java:
```diff
@@ -0,0 +1,17 @@
+import java.time.LocalDateTime;
+
+@Builder
+public record DonationCreateDto(
```
![image](https://github.com/user-attachments/assets/09a16ef9-1099-48c4-8e32-9985031ce0be)

salikhdev committed:
Donation is fixed (For cross review no test) 1.1v

# Conflicts:
#	src/main/java/faang/school/projectservice/controller/GlobalExceptionHandler.java

az3l1t reviewed (approved):
круто, вопросов нет

fladx commented on src/main/java/faang/school/projectservice/mapper/DonationMapper.java:
```diff
@@ -0,0 +1,17 @@
+import org.mapstruct.Mapper;
+import org.mapstruct.Mapping;
+
+@Mapper
+public interface DonationMapper {
+
+    @Mapping(target = "campaign.id", source = "campaignId")
+    Donation toEntity(DonationCreateDto dto);
+
+    @Mapping(target = "campaignId", source = "campaign.id")
+    DonationDto toDto(Donation donation);
+}
```
а ты тестил код? спринг инжектит этот маппер?

fladx commented:
Почему без тестов то) сделай тесты

salikhdev committed:
Donation add testing 1.2v

SnurnicynKonstantin reviewed (changes requested):
Приложение не запускается. Что-то с маппером, надо разобраться
<img width="1450" alt="image" src="https://github.com/user-attachments/assets/a903ed01-a756-4320-b123-e5b31f8467f4" />

salikhdev committed:
Donation add testing 1.3v

SnurnicynKonstantin commented on src/main/java/faang/school/projectservice/service/DonationService.java:
```diff
@@ -0,0 +1,118 @@
+        return donationMapper.toDto(donation);
+    }
+
+    protected PaymentResponse paymentToDonate(DonationCreateDto donationDto) {
```
Почему не приват?
И публичные методы сначала вверху идут, а приватные и протектед в конце класса

salikhdev replied:
> Почему не приват? И публичные методы сначала вверху идут, а приватные и протектед в конце класса

тогда не могу тестировать этого методу

SnurnicynKonstantin replied:
Ну и ладно. Он же используется в других методах, значит влияет на них. Вот в тестах для других методов проверяй, что эта функция отдает, что жидаешь, что клиент вызывался например, или что клиент ошибку выдал, тогда и вызываемый метод должен реагировать.
Мы пишем не код для тестов, а тесты для кода. По этому плохая практика приватные методы открывать  для того, что в тестах их удобно было прогнать

SnurnicynKonstantin commented on src/main/java/faang/school/projectservice/service/DonationService.java:
```diff
@@ -0,0 +1,118 @@
+        }
+    }
+
+    public UserDto getUserById(Long userId) {
```
Используется только в этом классе. Лучше его сделать приватным

salikhdev replied:
> Используется только в этом классе. Лучше его сделать приватным

тогда не могу тестировать этого методу

SnurnicynKonstantin commented on src/main/java/faang/school/projectservice/util/RandomGenerator.java:
```diff
@@ -0,0 +1,16 @@
+import java.util.Random;
+
+@Component
+public class RandomGenerator {
```
Этот класс используется только один раз и в одном сервисе. Есть ли смысл выносить две строчки кода в отдельный класс?
Обычно такое делаем, когда есть общий код, который надо использовать в нескольких классах. Тут ситуация противоположная

SnurnicynKonstantin reviewed (changes requested):
По логике и тестам все хорошо

salikhdev committed:
Donation add testing 1.4v

SnurnicynKonstantin reviewed (changes requested):
Оставил один коммеент. 
И посмотри плиз, у тебя два теста упало
<img width="914" alt="image" src="https://github.com/user-attachments/assets/6e967a40-b068-43b2-8f9d-6edadffe949c" />

salikhdev committed:
Donation add testing 1.5v

SnurnicynKonstantin reviewed (approved):
Вот теперь все красиво. 
И клиентом воспользовался и тесты по максимуму описал, молодец.
